File: packages/core/src/evalrank_core/aggregation_identity.py
"""Portable identities for one immutable EvalRank aggregation input."""

from __future__ import annotations

import re
from typing import Any

from evalrank_core.canonical_json import MAX_SAFE_INTEGER, canonical_json, sha256_hex


_DIGEST_RE = re.compile(r"^[0-9a-f]{64}$")
_CALIBRATION_REPORT_ID_RE = re.compile(r"^calibration_[0-9a-f]{64}$")
_OBSERVATION_ID_RE = re.compile(r"^obs_[0-9a-f]{64}$")


def aggregation_input_document(
    *,
    admission_cohort_digest: str,
    calibration_report_id: str,
    methodology_version: str,
    observation_ids: list[str],
    ranking_group: list[str],
) -> dict[str, Any]:
    """Return the exact validated restricted-JCS aggregation preimage."""

    observations = _observation_ids(observation_ids)
    group = _ranking_group(ranking_group)
    document = {
        "admission_cohort_digest": _pattern_string(
            "admission_cohort_digest",
            admission_cohort_digest,
            _DIGEST_RE,
        ),
        "calibration_report_id": _pattern_string(
            "calibration_report_id",
            calibration_report_id,
            _CALIBRATION_REPORT_ID_RE,
        ),
        "methodology_version": _nonempty_string(
            "methodology_version",
            methodology_version,
        ),
        "observation_ids": observations,
        "ranking_group": group,
    }
    canonical_json(document)
    return document


def derive_aggregation_input_digest(
    *,
    admission_cohort_digest: str,
    calibration_report_id: str,
    methodology_version: str,
    observation_ids: list[str],
    ranking_group: list[str],
) -> str:
    """Hash the exact aggregation input document as lowercase SHA-256."""

    return sha256_hex(
        aggregation_input_document(
            admission_cohort_digest=admission_cohort_digest,
            calibration_report_id=calibration_report_id,
            methodology_version=methodology_version,
            observation_ids=observation_ids,
            ranking_group=ranking_group,
        )
    )


def bootstrap_seed_document(
    aggregation_input_digest: str,
    methodology_version: str,
) -> dict[str, str]:
    """Return the exact validated restricted-JCS bootstrap-seed preimage."""

    document = {
        "aggregation_input_digest": _pattern_string(
            "aggregation_input_digest",
            aggregation_input_digest,
            _DIGEST_RE,
        ),
        "methodology_version": _nonempty_string(
            "methodology_version",
            methodology_version,
        ),
    }
    canonical_json(document)
    return document


def derive_bootstrap_seed(
    aggregation_input_digest: str,
    methodology_version: str,
) -> int:
    """Derive the portable safe-integer seed from the first SHA-256 bytes."""

    digest = sha256_hex(
        bootstrap_seed_document(aggregation_input_digest, methodology_version)
    )
    return int(digest[:16], 16) & MAX_SAFE_INTEGER


def _observation_ids(value: Any) -> list[str]:
    if not isinstance(value, list):
        raise TypeError("observation_ids must be an array")
    if not value:
        raise ValueError("observation_ids must be non-empty")
    observations = [
        _pattern_string("observation_ids item", item, _OBSERVATION_ID_RE)
        for item in value
    ]
    if len(observations) != len(set(observations)):
        raise ValueError("observation_ids must be unique")
    return sorted(observations)


def _ranking_group(value: Any) -> list[str]:
    if not isinstance(value, list):
        raise TypeError("ranking_group must be an array")
    if len(value) != 4:
        raise ValueError("ranking_group must contain exactly four strings")
    return [
        _nonempty_string(f"ranking_group[{index}]", item)
        for index, item in enumerate(value)
    ]


def _pattern_string(name: str, value: Any, pattern: re.Pattern[str]) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{name} must be a string")
    if pattern.fullmatch(value) is None:
        raise ValueError(f"{name} has an invalid format")
    return value


def _nonempty_string(name: str, value: Any) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{name} must be a string")
    if not value:
        raise ValueError(f"{name} must be non-empty")
    return value


__all__ = [
    "aggregation_input_document",
    "bootstrap_seed_document",
    "derive_aggregation_input_digest",
    "derive_bootstrap_seed",
]

File: packages/sdk-ts/src/aggregation-identity.ts
/** Portable identities for one immutable EvalRank aggregation input. */

import {
  MAX_SAFE_INTEGER,
  canonicalJson,
  sha256Hex,
} from "./decision-contracts.ts";


const digestPattern = /^[0-9a-f]{64}$/;
const calibrationReportIdPattern = /^calibration_[0-9a-f]{64}$/;
const observationIdPattern = /^obs_[0-9a-f]{64}$/;
const aggregationDocumentKeys = [
  "admission_cohort_digest",
  "calibration_report_id",
  "methodology_version",
  "observation_ids",
  "ranking_group",
] as const;

export type RankingGroupIdentity = readonly [string, string, string, string];

export interface AggregationInputDocument {
  admission_cohort_digest: string;
  calibration_report_id: string;
  methodology_version: string;
  observation_ids: string[];
  ranking_group: RankingGroupIdentity;
}

export interface BootstrapSeedDocument {
  aggregation_input_digest: string;
  methodology_version: string;
}

export function aggregationInputDocument(value: unknown): AggregationInputDocument {
  const payload = exactAggregationDocument(value);
  const document: AggregationInputDocument = {
    admission_cohort_digest: patternString(
      payload.admission_cohort_digest,
      digestPattern,
      "admission_cohort_digest",
    ),
    calibration_report_id: patternString(
      payload.calibration_report_id,
      calibrationReportIdPattern,
      "calibration_report_id",
    ),
    methodology_version: nonEmptyString(
      payload.methodology_version,
      "methodology_version",
    ),
    observation_ids: observationIds(payload.observation_ids),
    ranking_group: rankingGroup(payload.ranking_group),
  };
  canonicalJson(document);
  return document;
}

export async function deriveAggregationInputDigest(value: unknown): Promise<string> {
  return sha256Hex(aggregationInputDocument(value));
}

export function bootstrapSeedDocument(
  aggregationInputDigest: unknown,
  methodologyVersion: unknown,
): BootstrapSeedDocument {
  const document = {
    aggregation_input_digest: patternString(
      aggregationInputDigest,
      digestPattern,
      "aggregation_input_digest",
    ),
    methodology_version: nonEmptyString(methodologyVersion, "methodology_version"),
  };
  canonicalJson(document);
  return document;
}

export async function deriveBootstrapSeed(
  aggregationInputDigest: unknown,
  methodologyVersion: unknown,
): Promise<number> {
  const digest = await sha256Hex(
    bootstrapSeedDocument(aggregationInputDigest, methodologyVersion),
  );
  const firstEightBytes = BigInt(`0x${digest.slice(0, 16)}`);
  return Number(firstEightBytes & BigInt(MAX_SAFE_INTEGER));
}

function exactAggregationDocument(value: unknown): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new TypeError("aggregation input must be an object");
  }
  const payload = value as Record<string, unknown>;
  const actualKeys = Object.keys(payload).sort();
  const expectedKeys = [...aggregationDocumentKeys].sort();
  if (
    actualKeys.length !== expectedKeys.length
    || actualKeys.some((key, index) => key !== expectedKeys[index])
  ) {
    throw new TypeError("aggregation input must contain exactly the documented fields");
  }
  return payload;
}

function observationIds(value: unknown): string[] {
  if (!Array.isArray(value)) {
    throw new TypeError("observation_ids must be an array");
  }
  if (value.length === 0) {
    throw new TypeError("observation_ids must be non-empty");
  }
  for (let index = 0; index < value.length; index += 1) {
    if (!(index in value)) throw new TypeError("observation_ids must not be sparse");
  }
  const observations = value.map((item) =>
    patternString(item, observationIdPattern, "observation_ids item")
  );
  if (new Set(observations).size !== observations.length) {
    throw new TypeError("observation_ids must be unique");
  }
  return [...observations].sort();
}

function rankingGroup(value: unknown): RankingGroupIdentity {
  if (!Array.isArray(value)) {
    throw new TypeError("ranking_group must be an array");
  }
  if (value.length !== 4) {
    throw new TypeError("ranking_group must contain exactly four strings");
  }
  for (let index = 0; index < value.length; index += 1) {
    if (!(index in value)) throw new TypeError("ranking_group must not be sparse");
  }
  const group = value.map((item, index) =>
    nonEmptyString(item, `ranking_group[${index}]`)
  );
  return group as [string, string, string, string];
}

function patternString(value: unknown, pattern: RegExp, name: string): string {
  if (typeof value !== "string") {
    throw new TypeError(`${name} must be a string`);
  }
  if (!pattern.test(value)) {
    throw new TypeError(`${name} has an invalid format`);
  }
  return value;
}

function nonEmptyString(value: unknown, name: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new TypeError(`${name} must be a non-empty string`);
  }
  return value;
}
